### PR TITLE
Proper env for obj locals in comprehensions

### PR DESCRIPTION
**Pull request by sbarzowski**
We didn't set the environment (upvalues) for objects
created as comprehensions - we set them for each field
separately, but that meant missing the locals.

### CODE REVIEW

**Comment by paulfantom** [comment]
I can confirm that runtime error is gone. :tada: However output is different than one from C++ version.

I was comparing it by running both jsonnet binaries in https://github.com/openshift/cluster-monitoring-operator repo by executing `hack/build-jsonnet.sh` which is internally running `jsonnet -J jsonnet/vendor jsonnet/main.jsonnet > tmp/main.json` and copying result files.

1. with google/jsonnet in version v0.12.1 I don't get any changes in that repo:
```sh
$ ./hack/build-jsonnet.sh >/dev/null 2>&1
$ git status
On branch master
Your branch is up to date with 'origin/master'.

nothing to commit, working tree clean
```
2. with binary from this PR I get changed files:
```sh
$ ./hack/build-jsonnet.sh >/dev/null 2>&1
$ git status
On branch master
Your branch is up to date with 'origin/master'.

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git checkout -- <file>..." to discard changes in working directory)

        modified:   assets/alertmanager/secret.yaml
        modified:   assets/grafana/dashboard-definitions.yaml
        modified:   assets/prometheus-k8s/kube-rbac-proxy-secret.yaml

no changes added to commit (use "git add" and/or "git commit -a")
```
output of `git diff` can be found here: https://pastebin.com/raw/F5Gi0Wfb

**Comment by sbarzowski** [comment]
Ok, I found the issue that caused the difference and updated the PR. The object locals did field lookups from the wrong inheritance level, so combining super with object locals could sometimes lead to incorrect results.

There is still a difference in `assets/alertmanager/secret.yaml` and `assets/prometheus-k8s/kube-rbac-proxy-secret.yaml`, but it seems to be alright (there is no longer a trailing space in base64 encoded yaml).

@paulfantom Could you confirm if it's alright now?

File: builtins.go [hunk 1]
```diff
@@ -853,14 +852,23 @@ func builtinUglyObjectFlatMerge(i *interpreter, trace TraceElement, x value) (va
...
-			// we assume that the locals are the same for each of merged objects
-			locals = simpleObj.locals
 		}
+		anyObj = simpleObj
+	}
+
+	var locals []objectLocal
+	var localUpValues bindingFrame
+	if len(objarr.elements) > 0 {
+		simpleObj := anyObj
```
Comment (sparkprime) [inline builtins.go]: could delete this line and just use anyObj

**Comment by sparkprime** [comment]
LGTM

File: builtins.go (hunk 1 above)
Comment (sbarzowski) [reply builtins.go]: Thanks, fixed

**Comment by redbaron** [comment]
For my education,  is there an simplified example of which shows what changed? (link to slack would be fine)

**Comment by sbarzowski** [comment]
@redbaron It wasn't discussed on Slack, only here and in #274. The bugs were there only for like a week, so if you didn't upgrade during that time you shouldn't be affected.

The added tests could serve as simplified examples, in particular one bug was appearing on this 
https://github.com/google/go-jsonnet/pull/275/files#diff-c33488f5c65d3df5205c8ccdc3fc8060
It would actually print `{ bar: 4 }` - fixed by "Perform lookups from obj locals with right sb".

The other bug was the crash on tests such as these: https://github.com/google/go-jsonnet/pull/275/files#diff-4efdd34cd23b5169b125c582ed246b33
The environment passed to object locals in object comprehensions was always empty.

**Comment by redbaron** [comment]
> The bugs were there only for like a week, so if you didn't upgrade during that time you shouldn't be affected.

Nice! Good that it was caught so early. Thanks for clarification

**Pull request merged**